### DISCUSSION TITLE
How to define variable in layout, and setting the variable in page

### DISCUSSION THREAD

**Discussion by desbear** [joomlatools/joomlatools-pages#885] [Q&A] [answered]
I give a variable in my layout.
`<main class="<?= $pageclass?>">`
Then I create a page use this layout, define the variable in the page.
`<? $pageclass => 'someclass'?>`
but the output of the page didn't show the class name.

**Comment by johanjanssens** [comment]
Variables you define in a page, or a layout only exist in the context of that layout or page. This is done to prevent collisions between variables names. To pass a variable from a page to the layout you do the following:

### In your page

To pass the variables to the page you need to define them as part of the `@layout` attribute. By default its written as `@layout: [/path/to/file]` however you can also use the expanded notation to pass in additional properties, in that case you do the following:

```yaml
@layout:
    path: /default
    foo: bar 
    classes:
        a: b
        x: y
    callout: false
```

The `path` is the path of the layout and is a protected name, you cannot use that as a variable, all the rest you can choose freely. In this example I set 3 variables:

- `foo`
- `classes`
- `callout`

### In your layout

In your layout you can now just use these variables: 

```php
<?= $foo ?>
<?= $classes ?>
<?= $callout ?>
```

You can also make define defaults in your layout. Your code could then look like:

```yaml
---
@layout: /index
foo:  
classes: []
callout: true
---

<?= $foo ?>
<?= $classes ?>
<?= $callout ?>
```

Note that since 0.21 you can also define the defaults in partials: https://github.com/joomlatools/joomlatools-pages/pull/703. 

### Advanced

There also is a `layout()` function that gives you access to the layout properties at runtime. This function returns a [config object](https://github.com/joomlatools/joomlatools-pages/blob/master/code/object/config/config.php) and contains all properties and their merged values. 

It allows you to dynamically inspect the layout properties, add new properties and override existing ones, giving you full control. I do recommend to use the frontmatter approach as the default approach and use `layout()` only for more advanced requirements.

For more info: https://github.com/joomlatools/joomlatools-pages/pull/540

**Comment by desbear** [comment, marked as answer]
Than you ~ :)
I define the layout like this screenshot.
<img width="555" alt="截圖 2022-03-27 下午12 31 04" src="https://user-images.githubusercontent.com/545068/160266736-da668e3a-45c9-4a48-af9f-68b59dd0b583.png">
And this is the page
<img width="789" alt="截圖 2022-03-27 下午12 32 30" src="https://user-images.githubusercontent.com/545068/160266768-1bb35d50-39cf-4f79-8c1b-765dea5f5412.png">
but the output is <div class="wrapper default">
Did I do something wrong?

**Reply by johanjanssens** [reply]
In the second screenshot you have `layout` twice: layout and @layout. Pages is picking up the first one, removing it should solve it. Please note that this requires at least 0.21.

To learn about the meaning of the @ and the difference between a frontmatter attribute and a frontmatter property please read: https://github.com/joomlatools/joomlatools-pages/pull/700

**Reply by desbear** [reply]
Hi johanjanssens
thank U again.
<img width="1261" alt="截圖 2022-03-27 下午9 28 01" src="https://user-images.githubusercontent.com/545068/160283885-6d45fe5f-9c3a-4de2-a6de-f108307d8d45.png">
I remove the first layout. but still not work.
my com_pages version is 0.22.4

**Reply by waseemsadiq** [reply]
@desbear 

In the `Layout` where you want to `render` the class name:
<img width="364" alt="Screenshot 2022-03-27 at 15 40 48" src="https://user-images.githubusercontent.com/3632629/160286818-e7b3470f-94fd-4fee-8f43-54fdf8e03944.png">


In the `Page` where you want to `set` the class name:
![Screenshot 2022-03-27 at 15 17 36](https://user-images.githubusercontent.com/3632629/160285905-16440ac1-14a7-4ed7-a26e-913d81f28eb6.png)

Result:
![Screenshot 2022-03-27 at 15 20 08](https://user-images.githubusercontent.com/3632629/160285951-6b4d548e-2fb5-4588-a5cb-fdd5eb475962.png)

**Reply by desbear** [reply]
thank you~! it's work~!
